### PR TITLE
Improve fallback key behaviour

### PR DESCRIPTION
**Pull request by uhoreg**
- tracks whether the fallback key has already been published
- clear out old fallback key some time after it's been replaced

~Note: do not merge until libolm 3.2.7 is released.  (Also, needs a change to `package.json` to use the newer libolm) (test failures are due to this)~ Done

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improve fallback key behaviour ([\#2037](https://github.com/matrix-org/matrix-js-sdk/pull/2037)).

### CODE REVIEW

File: src/crypto/index.ts [hunk 1]
```diff
@@ -1864,8 +1865,23 @@ export class Crypto extends EventEmitter {
...
 
                 if (this.getNeedsNewFallback()) {
-                    logger.info("generating fallback key");
-                    await this.olmDevice.generateFallbackKey();
+                    const fallbackKeys = await this.olmDevice.getFallbackKey();
+                    // if fallbackKeys is non-empty, we've already generated a
+                    // fallback key, but it hasn't been published yet, so we
+                    // can use that instead of generating a new one
+                    if (!fallbackKeys.curve25519 ||
+                        [...Object.keys(fallbackKeys.curve25519)].length == 0) {
```
Comment (turt2live) [inline src/crypto/index.ts:1873]:
```suggestion
                        Object.keys(fallbackKeys.curve25519).length == 0) {
```

**Review by turt2live — APPROVED** [review]
yay, improved fallback keys

File: src/crypto/index.ts (hunk 1 above)
Comment (uhoreg) [reply src/crypto/index.ts:1873]: Huh. I was under the impression that `Object.keys` returns an iterator rather than an array, but looks like I was wrong.

**Comment by uhoreg** [comment]
Apparently, https://github.com/matrix-org/matrix-js-sdk/pull/2037/commits/ae68f000435959d48e88aa925f8210d4e337a354 makes jest run for over an hour, which is probably not ideal.  Will try to figure out a workaround.

**Comment by t3chguy** [comment]
Look into jest.useFakeTimers()

**Comment by uhoreg** [comment]
I think my changes exceed the "needs re-review" threshold.  (I also added support for unprefixed properties, to match https://github.com/matrix-org/synapse/pull/11541)

**Review by turt2live — APPROVED** [review]
Thanks :)

**Pull request merged**